### PR TITLE
Cherry-pick: Update go to 1.21.4 (#648)

### PR DESCRIPTION
**Pull request by nckturner**
Cherry pick of https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/648

**What this PR does / why we need it**:
Update go to v1.21.4

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 ARG image=public.ecr.aws/eks-distro-build-tooling/eks-distro-minimal-base-nonroot:2023-02-22-1677092456.2
-ARG golang_image=public.ecr.aws/eks-distro-build-tooling/golang:1.19-gcc
+ARG golang_image=public.ecr.aws/docker/library/golang:1.21.4
```
Comment (micahhausler) [inline Dockerfile:15]: Same as #658, happy to swich if AWS builds override this

Comment (nckturner) [reply Dockerfile:15]: Yeah this was changed by @dims in master, asked for his perspective on it and will update this PR if necessary.

Comment (nckturner) [reply Dockerfile:15]: Probably makes sense to leave the build image alone in the release branches

Comment (nckturner) [reply Dockerfile:15]: Discussed with @dims and we will inject the eks-distro image internally and this repo will reference the docker/library/golang:1.21.4 image, so long as it doesn't effect the authenticator behavior in any way (it shouldn't).

**Pull request merged**